### PR TITLE
Custom databaker filter - cdid filter

### PR DESCRIPTION
**Pull request by osamede20**
https://github.com/GSS-Cogs/de-helpers/issues/1

cdid.py created to pick out CDIDs in transforms
__init_py updated to include cdid_filter library
test_cdid.py - basic unit test created to test cdid.py

### CODE REVIEW

**Comment by mikeAdamss** [comment]
@osamede20 - nice one! 

I don't have time to read this properly at the mo sorry (tomorrow or monday) but one thing did occur to me earlier, we might need to check types.

so what happens if someone passed say a float, a datetime, a class, a None a boolean or an integer to it?

_shouldn't_ happen, but the filter needs to handle it (False for anything that's not a str is fine...I think) and you'd need a test to throw a few different types at it, to test/confirm the result.

**Comment by osamede20** [comment]
> @osamede20 - nice one!
> 
> I don't have time to read this properly at the mo sorry (tomorrow or monday) but one thing did occur to me earlier, we might need to check types.
> 
> so what happens if someone passed say a float, a datetime, a class, a None a boolean or an integer to it?
> 
> _shouldn't_ happen, but the filter needs to handle it (False for anything that's not a str is fine...I think) and you'd need a test to throw a few different types at it, to test/confirm the result.

**Comment by osamede20** [comment]
@mikeAdamss 

I mistakenly closed the PR.

Thanks for the feedback. I did not think of the points you raised. I will effect the changes.

File: dehelpers/filters/cdid.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# ---
```
Comment (mikeAdamss) [inline dehelpers/filters/cdid.py]: you can get rid all these commented out lines, it's just jupyter kruft that's not relevant here.

File: dehelpers/filters/__init__.py [hunk 2]
```diff
@@ -1 +1,4 @@
 from dehelpers.filters.example import foofilter
```
Comment (mikeAdamss) [inline dehelpers/filters/__init__.py]: can you do me a favour and rename `foofilter` to `is_true` please, just to keep with our new standards.

File: dehelpers/filters/__init__.py [hunk 3]
```diff
@@ -1 +1,4 @@
...
+from dehelpers.filters.cdid import cdid_filter
```
Comment (mikeAdamss) [inline dehelpers/filters/__init__.py]: just needs a tweak to the new conventions, so `like_cdid`

File: dehelpers/filters/cdid.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+#       format_name: light
+#       format_version: '1.5'
+#       jupytext_version: 1.10.2
+#   kernelspec:
+#     display_name: Python 3
+#     language: python
+#     name: python3
+# ---
+
+def cdid_filter(xycell):
```
Comment (mikeAdamss) [inline dehelpers/filters/cdid.py]: `like_cdid`

File: tests/filters/test_cdid.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+# ---
```
Comment (mikeAdamss) [inline tests/filters/test_cdid.py]: comments can go

File: tests/filters/test_cdid.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+#     display_name: Python 3
+#     language: python
+#     name: python3
+# ---
+
+# +
+import unittest
+
+from tests.helpers import MockXYCell
+from dehelpers.filters import cdid_filter
```
Comment (mikeAdamss) [inline tests/filters/test_cdid.py]: `like_cdid`

File: tests/filters/test_cdid.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+        # Create some test data
+        mocked_cdid_cells = [
+            MockXYCell(value="MK2N", x=1, y=2),
+            MockXYCell(value="ABCD", x=1, y=3),
+            MockXYCell(value="XT3H", x=1, y=4),
+            MockXYCell(value="NIHN", x=1, y=5)
+        ]
+        for mock_cell in mocked_cdid_cells:
+            self.assertEqual(cdid_filter(mock_cell), True)
```
Comment (mikeAdamss) [inline tests/filters/test_cdid.py]: `like_cdid`

File: dehelpers/filters/cdid.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    if type(xycell.value)==str and xycell.value.isalnum() == True and xycell.value.isupper() == True and len(xycell.value) == 4:
```
Comment (mikeAdamss) [inline dehelpers/filters/cdid.py]: I'd maybe add something like:

```
if not isinstance(xycell.value, str):
   return False
   
if xycell.value.isalnum() == True and xycell.value.isupper() == True and len(xycell.value) == 4:
    etc ....
```
as it is, I _think_ (not 100% here) that if someone manages to pass in a xycell.value that doesn't have `isalnum()` , `isupper()` or `len()` methods you'll get an error.

not sure how you'd do that mind, but if we don't guard against it sooner or later someone will figure it out 😄

File: tests/filters/test_cdid.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            MockXYCell(value=True, x=3, y=6),
+            MockXYCell(value=19.6, x=3, y=7),
+            MockXYCell(value="", x=2, y=6),
+            MockXYCell(value=2019, x=2, y=7),
+            MockXYCell(value=None, x=2, y=8),
+            MockXYCell(value=2021-0o4-23, x=4, y=1)
+        ]
+
+        for mock_cell in mocked_non_cdid_cells:
+            self.assertEqual(cdid_filter(mock_cell), False)
```
Comment (mikeAdamss) [inline tests/filters/test_cdid.py]: `like_cdid`

**Review by mikeAdamss — COMMENTED** [review]
one actual comment, a bunch of things related to the conventions we agreed earlier and a request to tweak the example while you're in here (pretty please 😄 ). 

I do realise you haven't had a chance to go back to this since we settled the conventions btw, just had some down time to review.

**Comment by osamede20** [comment]
Thanks Mike!
I will effect the changes shortly.

**Comment by mikeAdamss** [comment]
closing, duplicate

**Pull request closed without merging**